File: Object Oriented Design/P6/imageCollage.cs
/* Nils Fenske
    3/9/2016
    Programming Assignment 6
    ImageCollage class v4

Class Invariants

This class of object holds a sequence of image id's

================================================================================

Interface Invariants

The user creates a collage by passing in the size and number of subscriptions paid for. 
The user has the ability to read the number of items in the sequence as well as the
number of subscriptions or display's available to them. The user also can add images 
by passing in an image id. The user can display images as well as replace images by 
knowing the index of the image they want to replace and passing in the new one. 

================================================================================

Implementation Invariants

Data members are protected to be used by subclasses. Image id's are stored in an 
array. The object is constructed in initial state with values at 0 and status at 
innactive until sequence is full. Displaying the images is kept track of in order to deactivate
after a certain number of uses (subscriptions). The user is responsible for knowing the
index of the image they wish to replace as well as passing in the new image. Number of images 
cannot be reduced. 

*/

namespace P6
{
    class imageCollage : Collage
    {
        protected bool active;
        protected int size;
        protected int count;
        protected int displays;
        protected int subscriptions;
        protected int replacements;

        protected int[] imageArray;

        // Preconditions: User is responsible for passing in size and subscriptions. 
        // Postconditions: Object set in initial innactive state.
        public imageCollage(int s, int c)
        {
            size = s;
            imageArray = new int[size];
            count = 0;
            displays = 0;
            subscriptions = c;
            replacements = 0;
        }

        // Preconditions: User passes in id of image. Cannot add more than size set by user.
        // Postconditions: Once object is full, object is set in active state and no more can be added.
        public virtual void addImage(int id)
        {
            if (count < size)
            {
                imageArray[count] = id;
                count++;
            }

            if (count == size) { active = true; }
        }

        // Preconditions: object must be active. Prints out the sequnce of images as well as number of subscriptions. 
        // Postconditions: Possibly deactivated if number of subscriptions reached. 
        public virtual int[] getSequence()
        {
            if (active)
            {
                if (displays < subscriptions)
                {
                    displays++;
                    return imageArray;
                }
            }
            if (displays == subscriptions) { active = false; }
            return null;
        }

        // Preconditions: Object must be active. User is respnsible for knowing index of image they wish to replace. Must pass in new image. 
        // Postconditions: Object still full and active with different sequence. 
        public virtual void replace(int index, int replacement)
        {
            if (active)
            {
                imageArray[index] = replacement;
                replacements++;
            }
        }

        public virtual string whoAmI()
        {
            return "imageCollage";
        }

        public int getSubs()
        {
            return subscriptions;
        }

        public int getCount()
        {
            return count;
        }
    }
}

File: Object Oriented Design/P4/imageCollage.cpp
/* Nils Fenske
2/22/2016
Programming Assignment 4
ImageCollage class v2

Implementation Invariants

Data members are protected to be used by subclasses.Image id's are stored in an
array.The object is constructed in initial state with values at 0 and status at
innactive until sequence is full.Displaying the images is kept track of in order to deactivate
after a certain number of uses(subscriptions).The user is responsible for knowing the
index of the image they wish to replace as well as passing in the new image.Number of images
cannot be reduced.

*/

#include "imageCollage.h"

int imageCollage::getCount()
{
	return _count;
}

int imageCollage::getSubscriptions()
{
	return _subscriptions;
}

imageCollage::imageCollage() {};

imageCollage::imageCollage(int quantity, int subs)
{
	size = quantity;
	imageArray = new int[size];
	_count = 0;
	displays = 0;
	_subscriptions = subs;
	replacements = 0;
	active = false;
}

imageCollage::~imageCollage()
{
	delete[] imageArray;
}

imageCollage::imageCollage(const imageCollage& x)
{
	if (x.active) 
	{
		active = x.active;
		size = x.size;
		imageArray = new int[size];
		_count = x._count;
		displays = x.displays;
		_subscriptions = x._subscriptions;
		replacements = x.replacements;

		for (int i = 0; i < size; i++) {
			imageArray[i] = x.imageArray[i];
		}
	}
}

void imageCollage::operator=(const imageCollage& rhs) 
{
	if (rhs.active && this != &rhs)
	{
		delete[] imageArray;
		active = rhs.active;
		size = rhs.size;
		imageArray = new int[size];
		_count = rhs._count;
		displays = rhs.displays;
		_subscriptions = rhs._subscriptions;
		replacements = rhs.replacements;

		for (int i = 0; i < size; i++) {
			imageArray[i] = rhs.imageArray[i];
		}
	}
	return;
}

void imageCollage::addImage(int id)
{
	if (_count < size)
	{
		imageArray[_count] = id;
		_count++;
	}

	if (_count == size) { active = true; }
}

void imageCollage::display()
{
	if (active)
	{
		cout << endl;
		cout << "Image Sequence: (";
		for (int i = 0; i <= size - 2; i++)
		{
			cout << imageArray[i] << ", ";
		}
		cout << imageArray[size - 1] << ")" << endl;
		cout << "Subscriptions: " << _subscriptions << endl;
		displays++;
		if (displays == _subscriptions) { active = false; }
	}
}

void imageCollage::replace(int index, int replacement)
{
	if (active)
	{
		imageArray[index] = replacement;
		replacements++;
	}
}

int* imageCollage::getSequence()
{
	if (active) { return imageArray; }
}
